[SDK-3028] Add advanced used cases for Swift (#5057)
braze-inc/braze-docs: _docs/_developer_guide/platform_integration_guides/swift/advanced_use_cases/linking.md

---
nav_title: Deep Linking
article_title: Deep Linking for iOS
platform: Swift
page_order: 0
description: "This article covers how to implement the universal deep linking delegate for your iOS app and examples on how to deep link to app settings."

> 

Reviewer: Can we update line 31 to use a capital `i` in `Info.plist`

</array>
```

> <<<<<<< HEAD

Reviewer 2: ```suggestion
```

### Step 2: Adding a scheme whitelist

You must declare the URL schemes you wish to pass to `canOpenURL(_:)` by adding the `LSApplicationQueriesSchemes` key to your app's Info.plist file. Attempting to call schemes outside this allowlist will cause the system to record an error in the device's logs, and the deep link will not open. An example of this error will look like this:

```
<Warning>: -canOpenURL: failed for URL: "yourapp://deeplink" – error: "This app is not allowed to query for scheme yourapp"
```

For example, if an in-app message should open the Facebook app when tapped, the app has to have the Facebook custom scheme (`fb`) in your allowlist. Otherwise, the system will reject the deep link. Deep links that direct to a page or view inside your own app still require that your app's custom scheme be listed in your app's `Info.plist`.

> Your example allowlist might look something like:

Reviewer 2: The example below should read `<string>fb</string>` for facebook. `fb://` works but there's no `facebook://` URL scheme in the Facebook app

> Starting with iOS 9, apps must have a whitelist of custom schemes that the app is allowed to open. Attempting to call schemes outside this list will cause the system to record an error in the device's logs, and the deep link will not open. An example of this error will look like this:

Reviewer: The Swift SDK supports iOS 11+, so we should remove any mention around previous versions (e.g. here with iOS 9) and assume all clients are on iOS 11+

Reviewer 3: The requirement to allowlist custom schemes is still true for iOS 11+, correct?

Author: That's correct

}
```

{% endtab %}
{% tab OBJECTIVE-C %}

```objc
- (BOOL)application:(UIApplication *)app openURL:(NSURL *)url options:(NSDictionary<UIApplicationOpenURLOptionsKey, id> *)options {
  NSString *path  = [url path];
  NSString *query = [url query];
  // Here you should insert code to take some action based upon the path and query.
  return YES;
}
```

{% endtab %}
{% endtabs %}

> 

Author: It's in the old obj-c docs, I removed them here since they don't seem to match the text

> To use universal links, make sure you have added a registered domain to your app's capabilities and have uploaded an `apple-app-site-association` file. Then implement the method `application:continueUserActivity:restorationHandler:` in your `AppDelegate`. For example:

Reviewer 3: I was surprised by the note that universal links weren't compatible with so many Braze features. Let me ask: What is the purpose of the ## Universal links section? Is it just to address how universal links would normally be implemented if the user did not care about Braze push notifications, IAM, etc. compatibility?

Reviewer 3: Based on your answer to the above, I might recommend moving the universal link warning to the top of this article, then calling out how the alternative implementations listed in this article allow users to accomplish deep links without universal links. 

I can help with this, just let me know!

Author: Yeah, these are still accurate. Multiple customers still use universal links for various reasons like other partners so this section is also mainly to call out how this interacts with our SDK.

}
```

{% endtab %}
{% endtabs %}

> 

Reviewer: In line 138, I think we should word it like:
```
Refer to [Apple's documentation][11] for more information.
```

> ATS is applied by default on iOS 9+. It requires that all connections use HTTPS and are encrypted using TLS 1.2 with forward secrecy. Refer to [Requirements for Connecting Using ATS][14] for more information. All images served by Braze to end devices are handled by a content delivery network ("CDN") that supports TLS 1.2 and is compatible with ATS.

Reviewer: Same note around `iOS 9` in line 145

Reviewer 3: @hokstuff For clarity, this means that the ATS workarounds described here would be necessary for all of our users? Have I understood you correctly there?

Unless they are specified as exceptions in your application's `Info.plist`, connections that do not follow these requirements will fail with errors that look something like this:

> ```

Reviewer 3: Similarly, are all of these options for partially or completely disability ATS something we still feel high confidence in endorsing/linking to?

## URL encoding

> The SDK percent-encodes links to create valid `NSURL`s. All link characters that are not allowed in a properly formed URL, such as Unicode characters, will be percent escaped.

Reviewer 2: ```suggestion
The SDK percent-encodes links to create valid `URL`s. All link characters that are not allowed in a properly formed URL, such as Unicode characters, will be percent escaped.
```

Using Swift types names.

> To decode an encoded link, use the `NSString` method [`stringByRemovingPercentEncoding`][8]. You must also return `true` in the `BrazeDelegate.braze(_:shouldOpenURL:)`. A call to action is required to trigger the handling of the URL by your app. 

Reviewer 2: ```suggestion
To decode an encoded link, use the `String` property [`removingPercentEncoding`][8]. You must also return `true` in the `BrazeDelegate.braze(_:shouldOpenURL:)`. A call to action is required to trigger the handling of the URL by your app. 
```

The documentation link is unchanged

You can declare a category for, or directly modify, the `Braze.WebViewController` class to apply customization to the web view. Check the class' [documentation][6] detail.

### Linking handling customization

The `BrazeDelegate` protocol can be used to customize the handling of URLs such as deep links, web URLs, and universal links. To set the delegate during Braze initialization, set a delegate object on the `Braze` instance. Braze will then call your delegate's implementation of `shouldOpenURL` before handling any URIs.

#### Integration example: BrazeDelegate

{% tabs %}
{% tab swift %}

```swift
func braze(_ braze: Braze, shouldOpenURL context: Braze.URLContext) -> Bool {
  if context.url.host == "MY-DOMAIN.com" {
    // Custom handle link here
    return true
  }
  // Let Braze handle links otherwise
  return false
}
```

{% endtab %}
{% tab OBJECTIVE-C %}

```objc
- (BOOL)braze:(Braze *)braze shouldOpenURL:(BRZURLContext *)context {
  if ([[context.url.host lowercaseString] isEqualToString:@"MY-DOMAIN.com"]) {
    // Custom handle link here
    return YES;
  }
  // Let Braze handle links otherwise
  return NO;
}
```

> 

Reviewer 3: @nigelfaustino There's a link to ABKURLDelegate.h on line 191. Was this included in error?

Author: Yes, removing that.

  }
```

{% endtab %}
{% tab OBJECTIVE-C %}

```objc
- (BOOL)application:(UIApplication *)application openURL:(NSURL *)url options:(NSDictionary<NSString *, id> *)options {

>   NSString *urlString = url.absoluteString.stringByRemovingPercentEncoding;

Reviewer 2: ```suggestion
  NSString *urlString = [url.absoluteString stringByRemovingPercentEncoding];
```

  // Handle urlString
  return YES;
}
```

{% endtab %}
{% endtabs %}

## Customization {#linking-customization}

### Default WebView customization

> The open-source `Braze.WebViewController` class displays web URLs opened by the SDK, typically when "Open Web URL Inside App" is selected for a web deep link.

Reviewer 2: ```suggestion
The `Braze.WebViewController` class displays web URLs opened by the SDK, typically when "Open Web URL Inside App" is selected for a web deep link.
```

> You can declare a category for, or directly modify, the `Braze.WebViewController` class to apply customization to the web view. Check the class's [documentation][6] for details.

Reviewer 2: ```suggestion
You can customize the `Braze.WebViewController` via the [`BrazeDelegate.braze(_:willPresentModalWithContext:)`](https://braze-inc.github.io/braze-swift-sdk/documentation/brazekit/brazedelegate/braze(_:willpresentmodalwithcontext:)-12sqy/) delegate method.
```

### Linking handling customization

The `BrazeDelegate` protocol can be used to customize the handling of URLs such as deep links, web URLs, and universal links. To set the delegate during Braze initialization, set a delegate object on the `Braze` instance. Braze will then call your delegate's implementation of `shouldOpenURL` before handling any URIs.

#### Integration example: BrazeDelegate

{% tabs %}
{% tab swift %}

```swift
func braze(_ braze: Braze, shouldOpenURL context: Braze.URLContext) -> Bool {
  if context.url.host == "MY-DOMAIN.com" {
    // Custom handle link here

>     return true

Reviewer 2: ```suggestion
    return false
```
One of the breaking change of the Swift [migration](https://braze-inc.github.io/braze-swift-sdk/documentation/braze/appboy-migration-guide#:~:text=MessageUI.delegate-,Warning,should%20be%20opened%20by%20Appboy%20(false)%20or%20not%20(true).,-SDK%20initialization%20and)

  }
  // Let Braze handle links otherwise

>   return false

Reviewer 2: ```suggestion
  return true
```

}
```

{% endtab %}
{% tab OBJECTIVE-C %}

```objc
- (BOOL)braze:(Braze *)braze shouldOpenURL:(BRZURLContext *)context {
  if ([[context.url.host lowercaseString] isEqualToString:@"MY-DOMAIN.com"]) {
    // Custom handle link here

>     return YES;

Reviewer 2: ```suggestion
    return NO;
```

  }
  // Let Braze handle links otherwise

>   return NO;

Reviewer 2: ```suggestion
  return YES;
```

}
```

{% endtab %}
{% endtabs %}

> <<<<<<< HEAD

Reviewer 2: ```suggestion
```

For more information, see [`BrazeDelegate`][23].

## Frequent use cases

> =======
> For an example, see [`ABKURLDelegate.h`][23].
> >>>>>>> bcad141d9c2dedc70fb0c45a8d74affc1288a581

Reviewer 2: ```suggestion
```

## Deep linking to app settings

You can take advantage of `UIApplicationOpenSettingsURLString` to deep link users to your app's settings from Braze's push notifications, in-app messages, and the News Feed.

To take users from your app into the iOS settings:
1. First, make sure your application is set up for either [scheme-based deep links][25] or [universal links][27].
2. Decide on a URI for deep linking to the **Settings** page (for example, `myapp://settings` or `https://www.braze.com/settings`).
3. If you are using custom scheme-based deep links, add the following code to your `application:openURL:options:` method:

{% tabs %}
{% tab swift %}

```swift
func application(_ app: UIApplication, open url: URL, options: [UIApplicationOpenURLOptionsKey : Any] = [:]) -> Bool {
  let path = url.path
  if (path == "settings") {

>     UIApplication.shared.openURL(URL(string:UIApplicationOpenSettingsURLString)!)

Reviewer 2: ```suggestion
    UIApplication.shared.openURL(URL(string:UIApplication.openSettingsURLString)!)
```

> 3. If you are using custom scheme-based deep links, add the following code to your `application:openURL:options:` method:

Reviewer 3: @nigelfaustino Line 299 says that universal links can be used to deep link to your app's settings, but earlier the tutorial said that universal links did not play nice with Braze features like push notifications or IAM. Is this claim on 299 accurate?
